### PR TITLE
v2 query subclient

### PR DESCRIPTION
**Pull request by tmahmood-microsoft**
## Checklist
- [ ] I have read the [contribution guidelines](https://github.com/Azure/azure-iot-sdk-csharp/blob/main/.github/CONTRIBUTING.md).
- [ ] I added or modified the existing tests to cover the change (we do not allow our test coverage to go down).
- [ ] This pull-request is submitted against the `main` branch.

### CODE REVIEW

File: iothub/service/src/IotHubServiceClient.cs [hunk 1]
```diff
@@ -141,6 +141,12 @@ public IotHubServiceClient(string hostName, AzureSasCredential credential, IotHu
         /// <seealso href="https://docs.microsoft.com/azure/iot-hub/iot-hub-automatic-device-management"/>
         public ConfigurationsClient Configurations { get; protected set; }
 
+        /// <summary>
+        /// Subclient of <see cref="IotHubServiceClient"/> for query.
+        /// Executes SQL query.
```
Comment (drwill-ms) [inline iothub/service/src/IotHubServiceClient.cs:146]:
```suggestion
        /// Subclient of <see cref="IotHubServiceClient"/> for executing queries using a SQL-like syntax.
```

File: iothub/service/src/IotHubServiceClient.cs [hunk 2]
```diff
@@ -141,6 +141,12 @@ public IotHubServiceClient(string hostName, AzureSasCredential credential, IotHu
...
+        /// </summary>
```
Comment (drwill-ms) [inline iothub/service/src/IotHubServiceClient.cs:147]:
```suggestion
        /// </summary>
        /// <seealso href="https://docs.microsoft.com/azure/iot-hub/iot-hub-devguide-query-language"/>
```

File: iothub/service/src/Query/QueryClient.cs [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+using System.Net.Http.Headers;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Azure.Devices.Common.Exceptions;
+using Microsoft.Azure.Devices.Http2;
+
+namespace Microsoft.Azure.Devices
+{
+    /// <summary>
+    /// Subclient of <see cref="IotHubServiceClient"/> for query. Executes SQL query.
```
Comment (drwill-ms) [inline iothub/service/src/Query/QueryClient.cs:14]: Repeat doc comment changes from the service client property for query client.

File: iothub/service/src/Query/QueryClient.cs [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+        /// </exception>
+        /// <exception cref="HttpRequestException">
+        /// If the HTTP request fails due to an underlying issue such as network connectivity, DNS failure, or server
+        /// certificate validation.
+        /// </exception>
+        /// <exception cref="OperationCanceledException">If the provided cancellation token has requested cancellation.</exception>
+        /// <seealso href="https://docs.microsoft.com/azure/iot-hub/iot-hub-automatic-device-management"/>
+        public virtual IQuery CreateQuery(string sqlQueryString, int? pageSize = null, CancellationToken cancellationToken = default)
+        {
+            Logging.Enter(this, $"Creating query", nameof(CreateQuery));
```
Comment (drwill-ms) [inline iothub/service/src/Query/QueryClient.cs:64]:
```suggestion
            if (Logging.IsEnabled)
                Enter(this, $"Creating query", nameof(CreateQuery));
```

Comment (drwill-ms) [reply iothub/service/src/Query/QueryClient.cs:64]: Add the enabled check for all logging calls.

File: iothub/service/src/Query/QueryClient.cs [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+            try
+            {
+                return new Query((token) => ExecuteAsync(
```
Comment (drwill-ms) [inline iothub/service/src/Query/QueryClient.cs:67]: When is this awaited?

File: iothub/service/src/Query/QueryClient.cs [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+        {
+            if (Logging.IsEnabled)
+                Logging.Enter(this, $"Creating query", nameof(CreateAsync));
+            try
+            {
+                return new Query(async (token) => await ExecuteAsync(
+                    sqlQueryString,
+                    pageSize,
+                    token,
+                    cancellationToken));
```
Comment (drwill-ms) [inline iothub/service/src/Query/QueryClient.cs:73]: ConfigureAwait(false)?

File: iothub/service/src/Query/QueryClient.cs [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+            }
+            catch (Exception ex)
+            {
+                Logging.Error(this, $"{nameof(CreateAsync)} threw an exception: {ex}", nameof(CreateAsync));
```
Comment (drwill-ms) [inline iothub/service/src/Query/QueryClient.cs:77]: Still elsewhere too.

```suggestion
                if (Logging.IsEnabled)
                    Logging.Error(this, $"{nameof(CreateAsync)} threw an exception: {ex}", nameof(CreateAsync));
```

File: iothub/service/src/Query/QueryClient.cs [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+            }
+            catch (Exception ex)
+            {
+                if (Logging.IsEnabled)
+                    Logging.Error(this, $"{nameof(CreateAsync)} threw an exception: {ex}", nameof(CreateAsync));
+                throw;
+            }
+            finally
+            {
+                Logging.Exit(this, $"Creating query", nameof(CreateAsync));
```
Comment (drwill-ms) [inline iothub/service/src/Query/QueryClient.cs:83]: You've missed other logging cases that need the `IsEnabled` check. Don't make your reviewers point out each instance. Check them all and fix them, please.

Comment (tmahmood-microsoft) [reply iothub/service/src/Query/QueryClient.cs]: Fixed it, originally added the change but forgot to save it before committing. 
Will keep that in mind for the future.

**Review by drwill-ms — APPROVED** [review]

**Pull request merged**